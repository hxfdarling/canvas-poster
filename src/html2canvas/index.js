/* eslint-disable
no-console
*/
/**
 *
 * 需要处理的场景
 * 背景色
 * 文本:颜色、换行、字体、样式线、省略号
 * 图片
 * 圆角
 * z-index：static,absolute,relative
 */
import { NodeParser } from './lib/NodeParser';

export default function drawDom(poster, dom) {
  const stack = NodeParser(dom);
  console.log(stack);
}
