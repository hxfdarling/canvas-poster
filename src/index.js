/* eslint-disable
no-param-reassign
 */

import createStyledText from './canvasText';

export function testx() {
  return 1;
}
export default class Poster {
  constructor(options) {
    this.options = Object.assign(
      {
        width: 200,
        height: 200,
        scale: 2,
      },
      options
    );
    this.canvas = document.createElement('canvas');
    this.canvas.width = this.options.width;
    this.canvas.height = this.options.height;
    this.context = this.canvas.getContext('2d');
  }
  getImageData() {
    return this.canvas.toDataURL('image/png');
  }
  drawImage(src, x, y, w, h) {
    const { context } = this;
    const { scale } = this.options;
    return new Promise((resolve, reject) => {
      const img = document.createElement('img');
      // This enables CORS
      img.crossOrigin = 'Anonymous';
      img.onload = () => {
        try {
          context.drawImage(img, x * scale, y * scale, w * scale, h * scale);
          resolve();
        } catch (e) {
          reject(e);
        }
      };
      img.onerror = () => {
        reject(new Error('image load error'));
      };
      img.src = src;
    });
  }
  drawTexts(texts, x = 0, y = 0, options = {}) {
    const { scale } = this.options;
    const { context } = this;
    createStyledText(context, texts, options).render(x * scale, y * scale);
  }
}
