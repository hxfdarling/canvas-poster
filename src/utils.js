export function loop() {}
export function loadImage(src) {
  return new Promise((resolve, reject) => {
    const img = document.createElement('img');
    // 在部分老机型上面，无法加载图片
    if (src.indexOf(';base64,') === -1) {
      // This enables CORS
      img.crossOrigin = 'Anonymous';
    }
    img.onload = () => resolve(img);

    img.onerror = () => {
      reject(new Error('image load error'));
    };
    img.src = src;
    if (img.complete === true) {
      // Inline XML images may fail to parse, throwing an Error later on
      setTimeout(() => {
        resolve(img);
      }, 500);
    }
  });
}
export const SMALL_IMAGE =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';
