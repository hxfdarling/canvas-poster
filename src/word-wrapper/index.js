/* eslint-disable
no-param-reassign,
no-plusplus
 */

// copy from https://github.com/mattdesl/word-wrapper/blob/master/index.js
// add feature to support break all

const newline = /\n/;
const newlineChar = '\n';
const whitespace = /\s/;

export default function(text, opt) {
  const lines = wordWrap(text, opt);
  return lines.map((line) => text.substring(line.start, line.end)).join('\n');
}

export function wordWrap(text, opt) {
  opt = opt || {};

  // zero width results in nothing visible
  if (opt.width === 0 && opt.mode !== 'nowrap') return [];

  text = text || '';
  const width = typeof opt.width === 'number' ? opt.width : Number.MAX_VALUE;
  const start = Math.max(0, opt.start || 0);
  const end = typeof opt.end === 'number' ? opt.end : text.length;
  const { mode } = opt;

  const measure = opt.measure || monospace;
  if (mode === 'pre') return pre(measure, text, start, end, width);
  return greedy(measure, text, start, end, width, mode);
}

function idxOf(text, chr, start, end) {
  const idx = text.indexOf(chr, start);
  if (idx === -1 || idx > end) return end;
  return idx;
}

function isWhitespace(chr) {
  return whitespace.test(chr);
}

function pre(measure, text, start, end, width) {
  const lines = [];
  let lineStart = start;
  for (let i = start; i < end && i < text.length; i++) {
    const chr = text.charAt(i);
    const isNewline = newline.test(chr);

    // If we've reached a newline, then step down a line
    // Or if we've reached the EOF
    if (isNewline || i === end - 1) {
      const lineEnd = isNewline ? i : i + 1;
      const measured = measure(text, lineStart, lineEnd, width);
      lines.push(measured);

      lineStart = i + 1;
    }
  }
  return lines;
}

function greedy(measure, text, start, end, width, mode) {
  // A greedy word wrapper based on LibGDX algorithm
  // https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFontCache.java
  const lines = [];

  let testWidth = width;
  // if 'nowrap' is specified, we only wrap on newline chars
  if (mode === 'nowrap') testWidth = Number.MAX_VALUE;
  const breakAll = mode === 'breakAll';
  while (start < end && start < text.length) {
    // get next newline position
    const newLine = idxOf(text, newlineChar, start, end);

    // eat whitespace at start of line
    while (start < newLine) {
      if (!isWhitespace(text.charAt(start))) break;
      start++;
    }

    // determine visible # of glyphs for the available width
    const measured = measure(text, start, newLine, testWidth);

    let lineEnd = start + (measured.end - measured.start);
    let nextStart = lineEnd + newlineChar.length;

    // if we had to cut the line before the next newline...
    if (lineEnd < newLine) {
      if (!breakAll) {
        // find char to break on
        while (lineEnd > start) {
          if (isWhitespace(text.charAt(lineEnd))) break;
          lineEnd--;
        }
      }
      if (lineEnd === start) {
        if (nextStart > start + newlineChar.length) nextStart--;
        // If no characters to break, show all.
        lineEnd = nextStart;
      } else {
        nextStart = lineEnd;
        // eat whitespace at end of line
        while (lineEnd > start) {
          if (!isWhitespace(text.charAt(lineEnd - newlineChar.length))) break;
          lineEnd--;
        }
      }
    }
    if (lineEnd >= start) {
      const result = measure(text, start, lineEnd, testWidth);
      lines.push(result);
    }
    start = nextStart;
  }
  return lines;
}

// determines the visible number of glyphs within a given width
function monospace(text, start, end, width) {
  const glyphs = Math.min(width, end - start);
  return {
    start,
    end: start + glyphs,
  };
}
