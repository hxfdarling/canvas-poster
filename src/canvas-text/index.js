/* eslint-disable
no-param-reassign
 */

import assign from 'object-assign';
import dprop from 'dprop';

import { wordWrap } from '../word-wrapper';

import renderText from './lib/render';

import {
  getBlockLength,
  computeAttributeGlyphs,
  setFontParams,
  indexOfAttribute,
  composeBuffer,
  getMaxAttrHeight,
} from './lib/util';

const baseSettings = {
  baseline: 'alphabetic',
  align: 'left',
  fillStyle: '#000',
  style: 'normal',
  variant: 'normal',
  weight: 'normal',
  family: 'sans-serif',
  size: 32,
};

export default function createStyledText(context, chunks, opts) {
  if (!context) {
    throw new TypeError(
      'must specify a CanvasRenderingContext2D as first parameter'
    );
  }

  let fullText = '';
  let maxLineWidth = 0;
  let height = 0;
  let data;
  let lines = [];
  let fonts = [];
  const defaultOpts = assign({}, opts);

  const stlyedText = {
    render,
    update,
    layout,
  };

  // some read-only values
  Object.defineProperties(stlyedText, {
    lines: dprop(() => lines),

    fonts: dprop(() => fonts),

    width: dprop(() => maxLineWidth),

    height: dprop(() => height),
  });

  update(chunks, opts);

  function update(newChunks, newOpts) {
    opts = assign({}, baseSettings, defaultOpts, newOpts);

    // accept array or single element for string data
    if (!Array.isArray(newChunks)) {
      newChunks = [newChunks || ''];
    }

    chunks = newChunks;

    // run an initial layout by default
    if (opts.layout !== false) {
      layout();
    }
  }

  function layout() {
    // copy data to avoid mutating user objects
    data = chunks.map((attrib) => {
      if (typeof attrib === 'string') {
        attrib = { text: attrib };
      }

      attrib = assign({}, opts, attrib);

      attrib.text = attrib.text || '';
      attrib.font = getFontStyle(attrib, opts);

      // approximate line height from pixel size
      if (typeof attrib.lineHeight === 'undefined') {
        attrib.lineHeight = getLineHeight(attrib.size);
      }

      return attrib;
    });

    fonts = data.map((attrib) => attrib.font);

    fullText = composeBuffer(data);
    lines = wordWrap(fullText, {
      width: opts.width,
      mode: opts.wordWrap,
      measure,
    });

    const lineSpacing = opts.lineSpacing || 0;
    for (let i = lines.length - 1; i >= 0; i--) {
      const line = lines[i];
      line.height = getMaxAttrHeight(data, line.start, line.end);
      line.height += lineSpacing;
    }

    maxLineWidth = lines.reduce((prev, line) => Math.max(line.width, prev), 0);
    // 居右处理
    if (opts.width) {
      maxLineWidth = Math.max(maxLineWidth, opts.width);
    }
    height = lines.reduce((prev, line) => prev + line.height, 0);
  }

  function render(originX, originY, renderFunc) {
    let cursorX = 0;
    let cursorY = lines.length > 0 ? lines[0].height : 0;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const { end } = line;
      let { start } = line;
      const lineWidth = line.width;
      const maxAttrHeight = line.height;

      while (start < end) {
        // find next attribute chunk for this string char
        const attribIdx = indexOfAttribute(data, start);
        if (attribIdx === -1) {
          break;
        }

        // for each attribute in this range ...
        const attrib = data[attribIdx];
        const length = getBlockLength(attrib, start, end);

        const text = fullText.substring(start, start + length);
        const { lineHeight } = attrib;

        let x = originX + cursorX;
        const y = originY + cursorY - lineHeight;
        if (opts.align === 'right') {
          x += maxLineWidth - lineWidth;
        } else if (opts.align === 'center') {
          x += (maxLineWidth - lineWidth) / 2;
        }

        setFontParams(context, attrib);
        const textWidth = context.measureText(text).width;
        cursorX += textWidth;

        if (renderFunc) {
          renderFunc(context, text, x, y, textWidth, lineHeight, attrib, i);
        } else {
          renderText(context, text, x, y, textWidth, lineHeight, attrib, i);
        }

        // skip to next chunk
        start += Math.max(1, length);
      }
      cursorX = 0;

      const next = lines[i + 1];
      if (next) {
        cursorY += next.height;
      } else {
        cursorY += maxAttrHeight;
      }
    }
  }

  function measure(text, start, end, width) {
    let availableGlyphs = 0;
    const first = start;
    let totalWidth = 0;

    while (start < end) {
      // find next attribute chunk for this string char
      const attribIdx = indexOfAttribute(data, start);
      if (attribIdx === -1) {
        break;
      }

      // for each attribute in this range ...
      const attrib = data[attribIdx];

      const result = computeAttributeGlyphs(
        context,
        fullText,
        totalWidth,
        attrib,
        start,
        end,
        width
      );
      availableGlyphs += result.available;
      totalWidth += result.width;

      // skip to next attribute
      start = attrib.index + attrib.text.length;
    }

    return {
      width: totalWidth,
      start: first,
      end: first + availableGlyphs,
    };
  }
  return stlyedText;
}

function getFontStyle(opt, defaults) {
  const style = opt.style || defaults.style;
  const variant = opt.variant || defaults.variant;
  const weight = String(opt.weight || defaults.weight);
  const family = opt.family || defaults.family;
  const fontSize = typeof opt.size === 'number' ? opt.size : defaults.size;
  const fontStyle = [style, variant, weight, `${fontSize}px`, family].join(' ');
  return fontStyle;
}

function getLineHeight(fontSize) {
  return fontSize + 1;
}
