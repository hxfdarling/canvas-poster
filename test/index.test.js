import { testx } from '../src';

describe('canvas-poster', () => {
  test('test', async () => {
    expect(testx()).toMatchSnapshot();
  });
  test('main', async () => {
    // const poster = new Poster();
    // poster.drawTexts([
    //   {
    //     text: 'test poster ',
    //     size: 30,
    //     fillStyle: '#510f0f',
    //   },
    // ]);
  });
});
