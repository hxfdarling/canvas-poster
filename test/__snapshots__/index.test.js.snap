// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`canvas-poster test 1`] = `1`;
